versioned_hdf5/subchunk_map.py

```diff
@@ -36,280 +52,405 @@
     )
 
 
-@cython.cfunc
-def _subindex_chunk_slice(
-    s_start: Py_ssize_t,
-    s_stop: Py_ssize_t,
-    s_step: Py_ssize_t,
-    c_start: Py_ssize_t,
-    c_stop: Py_ssize_t,
-) -> slice:
-    """Given a slice(s_start, s_stop, s_step) indexing an axis of
-    a dataset, return the slice of the output array (on __getitem__)
-    or value parameter array (on __setitem__) along the
-    same axis that is targeted by the same slice after it's been
-    clipped to select only the data within the range [c_start, c_stop[
-
-    In other words:
-
-    a = _subindex_chunk_slice(s_start, s_stop, s_step, c_start, c_stop)
-    b = _subindex_slice_chunk(s_start, s_stop, s_step, c_start, c_stop)
-
-    __getitem__: out[a] = cache_chunk[b]
-    __setitem__: cache_chunk[b] = value[a]
-    """
-    start: Py_ssize_t = max(c_start, s_start)
-    # Get the smallest lcm multiple of common that is >= start
-    start = smallest_step_after(start, s_start % s_step, s_step)
-    # Finally, we need to shift start so that it is relative to index
-    start = (start - s_start) // s_step
+class DropAxis(enum.Enum):
+    _drop_axis = 0
 
-    stop: Py_ssize_t = min(c_stop, s_stop)
-    stop = ceil_a_over_b(stop - s_start, s_step) if stop > s_start else 0
 
-    return slice(int(start), int(stop), 1)
+# Returned instead of an AnySlicer. Signals that the axis should be removed when
+# aggregated into an AnySlicerND.
+DROP_AXIS = DropAxis._drop_axis
 
 
-@cython.cfunc
-def _subindex_slice_chunk(
-    s_start: Py_ssize_t,
-    s_stop: Py_ssize_t,
-    s_step: Py_ssize_t,
-    c_start: Py_ssize_t,
-    c_stop: Py_ssize_t,
-) -> slice:
-    """Given a slice(s_start, s_stop, s_step) indexing an axis of
-    a dataset, return a slice that's been clipped to select only
-    the data within the range [c_start, c_stop[ and shifted back
-    by s_start.
-
-    See examples on _subindex_chunk_slice
+@cython.cclass
+class IndexChunkMapper:
+    """Abstract class that manipulates a numpy fancy index along a single axis of a
+    chunked array
+
+    Parameters
+    ----------
+    chunk_indices:
+        Array of indices of all the chunks involved in the selection along the axis
```

Comment:
For the purpose of `as_subchunk_map` exclusively, it would have been more efficient writing a function that returns an iterator of chunk indices. However, later PRs extensively rely on this being an actual array.